### ISSUE TITLE
ConditionalAttribute removes too much from the call site

### ISSUE THREAD

**Issue by Kryptos-FR** [dotnet/roslyn#28801]
**Version Used**: 
Tested in 2015 and 2017.
.Net Framework 4.5.2

**Steps to Reproduce**:
Consider the following extension method:
```csharp
[Conditional("DEBUG")]
public static void Forget(this Task task)
{
    if (task == null) throw new ArgumentNullException();
}
```
Used in a context of a async/await method:
```csharp
public async Task MethodAsync()
{
    // fire and forget task
    Task.Run(() => { }).Forget();
    // do something else
}
```

The goal of that specific extension method was to suppress the warning that the `Task.Run` was not awaited. But it could have other purposes as well.

**Expected Behavior**:
On the call to `Forget` is removed when compiled in release. The task is still run.

**Actual Behavior**:
The whole call site is removed. The code to run the task is not emitted in IL.

**Comparison of IL**:
ConditionalAttribute is "true:"
```IL
.method private hidebysig newslot virtual final 
        instance void  MoveNext() cil managed
{
  .override [mscorlib]System.Runtime.CompilerServices.IAsyncStateMachine::MoveNext
  // Code size       97 (0x61)
  .maxstack  2
  .locals init ([0] int32 V_0,
           [1] class [mscorlib]System.Exception V_1)
  IL_0000:  ldarg.0
  IL_0001:  ldfld      int32 ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>1__state'
  IL_0006:  stloc.0
  .try
  {
    IL_0007:  nop
    IL_0008:  ldsfld     class [mscorlib]System.Action ConditionalAttributeBug.Program/'<>c'::'<>9__1_0'
    IL_000d:  dup
    IL_000e:  brtrue.s   IL_0027
    IL_0010:  pop
    IL_0011:  ldsfld     class ConditionalAttributeBug.Program/'<>c' ConditionalAttributeBug.Program/'<>c'::'<>9'
    IL_0016:  ldftn      instance void ConditionalAttributeBug.Program/'<>c'::'<MethodAsync>b__1_0'()
    IL_001c:  newobj     instance void [mscorlib]System.Action::.ctor(object,
                                                                      native int)
    IL_0021:  dup
    IL_0022:  stsfld     class [mscorlib]System.Action ConditionalAttributeBug.Program/'<>c'::'<>9__1_0'
    IL_0027:  call       class [mscorlib]System.Threading.Tasks.Task [mscorlib]System.Threading.Tasks.Task::Run(class [mscorlib]System.Action)
    IL_002c:  call       void ConditionalAttributeBug.TaskExtensions::Forget(class [mscorlib]System.Threading.Tasks.Task)
    IL_0031:  nop
    IL_0032:  leave.s    IL_004c
  }  // end .try
  catch [mscorlib]System.Exception 
  {
    IL_0034:  stloc.1
    IL_0035:  ldarg.0
    IL_0036:  ldc.i4.s   -2
    IL_0038:  stfld      int32 ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>1__state'
    IL_003d:  ldarg.0
    IL_003e:  ldflda     valuetype [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>t__builder'
    IL_0043:  ldloc.1
    IL_0044:  call       instance void [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetException(class [mscorlib]System.Exception)
    IL_0049:  nop
    IL_004a:  leave.s    IL_0060
  }  // end handler
  IL_004c:  ldarg.0
  IL_004d:  ldc.i4.s   -2
  IL_004f:  stfld      int32 ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>1__state'
  IL_0054:  ldarg.0
  IL_0055:  ldflda     valuetype [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>t__builder'
  IL_005a:  call       instance void [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetResult()
  IL_005f:  nop
  IL_0060:  ret
} // end of method '<MethodAsync>d__1'::MoveNext
```

ConditionalAttribute is "false":
```IL
.method private hidebysig newslot virtual final 
        instance void  MoveNext() cil managed
{
  .override [mscorlib]System.Runtime.CompilerServices.IAsyncStateMachine::MoveNext
  // Code size       45 (0x2d)
  .maxstack  2
  .locals init ([0] class [mscorlib]System.Exception V_0)
  .try
  {
    IL_0000:  leave.s    IL_0019
  }  // end .try
  catch [mscorlib]System.Exception 
  {
    IL_0002:  stloc.0
    IL_0003:  ldarg.0
    IL_0004:  ldc.i4.s   -2
    IL_0006:  stfld      int32 ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>1__state'
    IL_000b:  ldarg.0
    IL_000c:  ldflda     valuetype [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>t__builder'
    IL_0011:  ldloc.0
    IL_0012:  call       instance void [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetException(class [mscorlib]System.Exception)
    IL_0017:  leave.s    IL_002c
  }  // end handler
  IL_0019:  ldarg.0
  IL_001a:  ldc.i4.s   -2
  IL_001c:  stfld      int32 ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>1__state'
  IL_0021:  ldarg.0
  IL_0022:  ldflda     valuetype [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>t__builder'
  IL_0027:  call       instance void [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetResult()
  IL_002c:  ret
} // end of method '<MethodAsync>d__1'::MoveNext
```

Expected IL:
```IL
.method private hidebysig newslot virtual final 
        instance void  MoveNext() cil managed
{
  .override [mscorlib]System.Runtime.CompilerServices.IAsyncStateMachine::MoveNext
  // Code size       82 (0x52)
  .maxstack  2
  .locals init ([0] class [mscorlib]System.Exception V_0)
  .try
  {
    IL_0000:  ldsfld     class [mscorlib]System.Action ConditionalAttributeBug.Program/'<>c'::'<>9__1_0'
    IL_0005:  dup
    IL_0006:  brtrue.s   IL_001f
    IL_0008:  pop
    IL_0009:  ldsfld     class ConditionalAttributeBug.Program/'<>c' ConditionalAttributeBug.Program/'<>c'::'<>9'
    IL_000e:  ldftn      instance void ConditionalAttributeBug.Program/'<>c'::'<MethodAsync>b__1_0'()
    IL_0014:  newobj     instance void [mscorlib]System.Action::.ctor(object,
                                                                      native int)
    IL_0019:  dup
    IL_001a:  stsfld     class [mscorlib]System.Action ConditionalAttributeBug.Program/'<>c'::'<>9__1_0'
    IL_001f:  call       class [mscorlib]System.Threading.Tasks.Task [mscorlib]System.Threading.Tasks.Task::Run(class [mscorlib]System.Action)
    IL_0024:  pop
    IL_0025:  leave.s    IL_003e
  }  // end .try
  catch [mscorlib]System.Exception 
  {
    IL_0027:  stloc.0
    IL_0028:  ldarg.0
    IL_0029:  ldc.i4.s   -2
    IL_002b:  stfld      int32 ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>1__state'
    IL_0030:  ldarg.0
    IL_0031:  ldflda     valuetype [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>t__builder'
    IL_0036:  ldloc.0
    IL_0037:  call       instance void [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetException(class [mscorlib]System.Exception)
    IL_003c:  leave.s    IL_0051
  }  // end handler
  IL_003e:  ldarg.0
  IL_003f:  ldc.i4.s   -2
  IL_0041:  stfld      int32 ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>1__state'
  IL_0046:  ldarg.0
  IL_0047:  ldflda     valuetype [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder ConditionalAttributeBug.Program/'<MethodAsync>d__1'::'<>t__builder'
  IL_004c:  call       instance void [mscorlib]System.Runtime.CompilerServices.AsyncTaskMethodBuilder::SetResult()
  IL_0051:  ret
} // end of method '<MethodAsync>d__1'::MoveNext
```

**Comment by CyrusNajmabadi** [comment]
>That also seems to break the rules of least astonishment.

I disagree.  it would be astonishing me to learn that calling an extension method would have different behavior if you called it as a static method or as if it was an instance.  in other words:

```ExtMethod(arg)``` and ```arg.ExtMethod()``` should always have the same semantics if it's the same method.  

Since ExtMethod has said that it's Conditional, i would not expect it to evaluate its arg.  That's the least-astonishing interpretation for me.

> And most code analyzer (ReSharper, Roslyn) will suggest to inline the task variable as it should not be a breaking change, but will be in that case.

Roslyn should not be suggesting that you inline the variable.  'inline variable' is a code-tool, so it's available as something you can invoke (on pretty much any variable).  But there should be no "suggestion" that you actually do this.

I cannot speak for resharper.  If it is doing this incorrectly, then that sounds like an issue they would have to fix.

**Comment by Kryptos-FR** [comment]
The problem is that it is not only for extension method. Consider this:

```csharp
class Program
{
    static void Main(string[] args)
    {
        new Fluent().Foo().Foo().Foo().Bar();
        new Fluent().Foo().Foo().Foo().Baz();
    }
}

class Fluent
{
    public Fluent Foo()
    {
        return this;
    }
        
    [Conditional("DEBUG")]
    public void Bar()
    {
    }
}

static class FluentExtensions
{
    [Conditional("DEBUG")]
    public static void Baz(this Fluent fluent)
    {
    }

}
```

What do you think happens when compiled in RELEASE? The whole code is skipped: main is empty.
Considering that fluent API are more and more common, that is wrong! Imagine if the `Conditional`-decorated method was just a debug method to dump something in the console. The rest of the code should still be called. Especially why would the previous `new Fluent().Foo().Foo()` calls also skipped. There are not supposed to be affected by the conditional attribute.

**Comment by CyrusNajmabadi** [comment]
> Especially the following code samples are supposed to be equivalent, but are not in practice:

This has never been true with conditional methods.   Just consider the following:

```c#
class Example
{
    public Example(int i)
    {
        Console.WriteLine(i);
    }

    [Conditional("release")]
    public void Foo()
    {
    }
}

class C
{
    static void Main(string[] args)
    {
        var ex = new Example(1);
        ex.Foo();
  
        new Example(2).Foo();
    }
}
```

Both expressions containing 'Foo' will be removed.  Leaving only ```var ex = new Example(1)```.  

When it comes to conditional methods, it's *never* been the case that inlining the receiver preserves semantics.  This is true even for instance methods.  It doesn't have to deal with extension methods at all :)

**Comment by Kryptos-FR** [comment]
How is that a correct behavior? That is definitely a design mistake.

**Comment by CyrusNajmabadi** [comment]
> What do you think happens when compiled in RELEASE? The whole code is skipped: main is empty

Yes.  That's my exact expectation.  Since Bar() is conditional and the condition does not apply, executing its receiver should not happen :)

> [they] are not supposed to be affected by the conditional attribute.

Yes they are.  The spec is clear here: https://github.com/dotnet/csharplang/blob/master/spec/attributes.md#the-conditional-attribute

> A method decorated with the Conditional attribute is a conditional method. The Conditional attribute indicates a condition by testing a conditional compilation symbol. Calls to a conditional method are either included or omitted depending on whether this symbol is defined at the point of the call. If the symbol is defined, the call is included; otherwise, the call (including evaluation of the receiver and parameters of the call) is omitted.



yes they are.

**Comment by Kryptos-FR** [comment]
Sorry but no, the docs doesn't state the the WHOLE call chain is removed. In my previous example, at least a few call to `Foo()` should have been preserved.

**Comment by CyrusNajmabadi** [comment]
> How is that a correct behavior?

The compiler is behaving exactly as specified.  That's def what i would call 'correct behavior' :)

> That is definitely a design mistake.

Design concerns should go to "dotnet/csharplang".  This is the repo for the Roslyn C# and VB compilers and IDEs.  Issues with the language itself go to the referenced repo :)

--

Now, i personally don't think this is a design mistake.  This has been the purpose and intent of ConditionalAttribute since inception.  It was entirley about not executing the receiver or parameters.

**Comment by Kryptos-FR** [comment]
> Design concerns should go to "dotnet/csharplang". This is the repo for the Roslyn C# and VB compilers and IDEs. Issues with the language itself go to the referenced repo :)

Conditional attribute affects the compiler. So definitely the correct repo :)

**Comment by CyrusNajmabadi** [comment]
> Sorry but no, the docs doesn't state the the WHOLE call chain is removed.

Yes it does.  That's what the "receiver" is.  The language defines an invocation as:

```
invocation_expression
    : primary_expression '(' argument_list? ')'
```

In this case, the entire "this.Foo().Foo().Foo" is the primary expression that defines the receiver.  The spec on Conditional attributes says the receiver will not be evaluated.  So this will not be evaluated :)

**Comment by CyrusNajmabadi** [comment]
> Conditional attribute affects the compiler. So definitely the correct repo :)

Your call.  But this will simply be closed as by-design in this repo.  When the compiler is behaving as the specification intends, it's not an issue with the compiler.  If you have an issue with the specification, then this is not the right repo.

**Comment by CyrusNajmabadi** [comment]
> At least improve the docs

Which docs are you referring specifically?  I ask, so the right people can be contacted.  Thanks.

> I find it surprising that the whole call chain would be considered a single receiver

This is how it works in most 'method-call' based languages.  Java, C++, C#, Go, Swift, Scala, etc. etc.

> That doesn't seem to fit what I was taught in language theory when I was a student.

I find that surprising.  Language theory almost always teaches that methods (at least instance ones) are a parameterized message being sent to some receiver instance.  

So clearly you need that receiver instance.   If the whole call chain was not that receiver... then what would hte method actually be invoked upon?  

> Especially introducing a variable to split the call chain should not have side-effects.

The same can be said about doing seomthing like passing a parameter.  Say you had this:

```c#
var v = Foo();
ConditionalMethod(v);
```

vs

```c#
ConditionalMethod(Foo());
```

You could argue that extracting an argument into a local (or vice-versa) should not have side-effects.  But that's exactly the point of ConditionalMethods.   You want to be able to have them in code and no that you won't have *anything* execute (receiver, arguments, nothing).

**Comment by CyrusNajmabadi** [comment]
> Especially introducing a variable to split the call chain should not have side-effects.

Sure it can.  Say you have this:

```
#if DEBUG
a().b().c().d();
#endif
```
And now you split that into:

```
var t = a().b();
#if DEBUG
t.c().d();
#endif
```

Clearly, we would expect that this wouldl behave differently.  Why?  Because part of it is *conditionally* compiled away.  That's the *exact* intent of the ConditionalAttribute.  It is a shorthand way to express the above.  It *conditionally* compiles away the evaluation of the reciever, the arguments, and the call to the method.

**Comment by CyrusNajmabadi** [comment]
Note: for your original code, you can always write this as:

```c#
public static void Forget(this Task task)
{
#ifdef DEBUG
    if (task == null) throw new ArgumentNullException();
#endif
}
```

This will ahve the semantics you want.  Note that the runtime will see that this method is empty in release mode and will entirely elide calls to it from the final code.

**Comment by Kryptos-FR** [comment]
In your example `DEBUG` affects all methods: `a`, `b, `c`, and `d`.
In case of a conditional attribute on `d` only, why should `a`, `b, `c` be affected by it?

**Comment by CyrusNajmabadi** [comment]
> In your example ```DEBUG``` affects all methods: ```a, b, c```, and ```d```. In case of a conditional attribute ondonly, why should ```a,b, c``` be affected by it?  

Because that's the "receiver" of the invocation.  The receiver is not evaluated.  That's the point :)   The purpose of ConditionalAttributes is so you can effectively get the same behavior as:

```c#
#if debug
<arbitrary receiver>.Meth(<arbitrary params>);
#endif
```

without actually having to surround with ```#if```s.

**Comment by CyrusNajmabadi** [comment]
Also, just as an FYI.  If you want to discuss questions/concerns with Roslyn, you can go to https://gitter.im/dotnet/roslyn to discuss things.  If you have questions/concerns with the language itself, then https://gitter.im/dotnet/csharplang is a good resource.

**Comment by jaredpar** [comment]
As indicated by @CyrusNajmabadi and @svick this behavior is By Design. The compiler is conforming to the language specification here and behaving as intended.

**Comment by svick** [comment]
@Kryptos-FR 

> At least improve the docs. Don't expect common developers to understand the jargon of the specs.

Nobody expects that. You claimed that the behavior is wrong, so it makes sense to go to the authoritative source to check whether that is the case. And that's what the spec is for.

As for documentation for "common developers", the only thing I found is [the API reference for `ConditionalAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.conditionalattribute). I do think that document should be improved, so I opened https://github.com/dotnet/docs/issues/6633.
